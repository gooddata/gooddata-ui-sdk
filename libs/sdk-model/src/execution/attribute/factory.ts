// (C) 2019-2023 GoodData Corporation
import identity from "lodash/identity.js";
import cloneDeep from "lodash/cloneDeep.js";
import isEmpty from "lodash/isEmpty.js";
import { IAttribute, isAttribute } from "./index.js";
import { ObjRef, objRefToString, Identifier, isObjRef } from "../../objRef/index.js";
import { idRef } from "../../objRef/factory.js";
import SparkMD5 from "spark-md5";
import { sanitizeLocalId } from "../../sanitizeLocalId.js";

/**
 * Input to the AttributeBuilder.
 * @public
 */
export type AttributeBuilderInput = Identifier | ObjRef | IAttribute;

/**
 * Builder for attributes.
 *
 * Do not instantiate this class directly. Instead use {@link newAttribute} or {@link modifyAttribute}.
 *
 * @public
 */
export class AttributeBuilder {
    private attribute: IAttribute["attribute"];
    private customLocalId = false;

    /**
     * @internal
     */
    constructor(input: AttributeBuilderInput) {
        if (isAttribute(input)) {
            this.attribute = cloneDeep(input.attribute);
            this.customLocalId = true;
        } else {
            const displayForm: ObjRef = isObjRef(input) ? input : idRef(input, "displayForm");

            this.attribute = {
                displayForm,
                localIdentifier: "",
            };
        }
    }

    /**
     * Sets alias - alternative title - for the attribute.
     *
     * @remarks
     * This value will then be used in various chart-specific descriptive elements. For convenience if no alias is specified,
     * the attribute will fall back to server-defined value.
     *
     * @param alias - alias to use instead of attribute title; undefined to use server-defined value
     */
    public alias = (alias?: string | undefined): this => {
        if (!alias) {
            return this.noAlias();
        }

        this.attribute.alias = alias;

        return this;
    };

    /**
     * Resets alias - alternative title - set for the attribute.
     *
     * @remarks
     * The server-defined title of the attribute will be used instead.
     */
    public noAlias = (): this => {
        delete this.attribute.alias;

        return this;
    };

    /**
     * Sets show all values property.
     *
     * @remarks
     * The flag showAllValues translates to a property of the same name on the attribute in execution definition.
     * If truthy, the backend will return all values of the particular attribute in the execution response
     * even if there are no data available for it.
     *
     * @param showAllValues - flag defining whether to return all attribute values for given attribute; undefined to use backend default behavior(false)
     */
    public showAllValues = (showAllValues?: boolean | undefined): this => {
        if (showAllValues === undefined) {
            delete this.attribute.showAllValues;
        } else {
            this.attribute.showAllValues = showAllValues;
        }

        return this;
    };

    /**
     * Sets display form reference.
     *
     * @param ref - new ref to use
     */
    public displayForm = (ref: ObjRef): this => {
        this.attribute.displayForm = ref;

        return this;
    };

    /**
     * Sets local identifier (localId) for the attribute. LocalId can be used to reference the attribute
     * within the execution definition.
     *
     * Normally, builder will generate localId based on contents of the attribute definition - taking all
     * properties into account: in typical scenarios you don't have to call this function at all. The only exception
     * where you have to provide custom local id is if your execution must contain the exact same attribute twice.
     *
     * For convenience, this method also accepts 'undefined', which indicates that the default local id generation
     * logic should be used.
     *
     * @param localId - local identifier to set; if not specified, the builder will ensure local id will
     * be generated
     */
    public localId = (localId?: Identifier | undefined): this => {
        if (!localId || localId.trim().length === 0) {
            return this.defaultLocalId();
        }

        this.attribute.localIdentifier = localId;
        this.customLocalId = true;

        return this;
    };

    /**
     * Indicates that the attribute's localId should be generated using the default local-id generator logic.
     */
    public defaultLocalId = (): this => {
        this.attribute.localIdentifier = "";
        this.customLocalId = false;

        return this;
    };

    /**
     * Creates the IAttribute instance.
     */
    public build = (): IAttribute => {
        const localIdentifier = this.getOrGenerateLocalId();

        return {
            attribute: {
                ...this.attribute,
                localIdentifier,
            },
        };
    };

    private getOrGenerateLocalId(): string {
        if (this.customLocalId && !isEmpty(this.attribute.localIdentifier)) {
            return this.attribute.localIdentifier;
        }

        return sanitizeLocalId(
            ["a", this.calculateAliasHash(), objRefToString(this.attribute.displayForm)]
                .filter((part) => !isEmpty(part))
                .join("_"),
        );
    }

    private calculateAliasHash(): string {
        if (!this.attribute.alias) {
            return "";
        }

        const hasher = new SparkMD5();
        hasher.append(this.attribute.alias);

        return hasher.end().substr(0, 8);
    }
}

/**
 * Function that will be called to perform modifications of an attribute before it is fully constructed.
 *
 * @public
 */
export type AttributeModifications = (builder: AttributeBuilder) => AttributeBuilder;

/**
 * Creates a new attribute with the specified display form ref and optional modifications and localIdentifier.
 * @param displayFormRefOrId - ref or identifier of the attribute display form
 * @param modifications - optional modifications (e.g. alias, etc.)
 * @public
 */
export function newAttribute(
    displayFormRefOrId: ObjRef | Identifier,
    modifications: AttributeModifications = identity,
): IAttribute {
    const builder = new AttributeBuilder(displayFormRefOrId);

    return modifications(builder).build();
}

/**
 * Allows modification of an existing attribute instance.
 *
 * @remarks
 * The returned attribute will have the same localId as the original attribute. If you would like to assign
 * new/different local identifier to the attribute, you can do that using the modifications where you can provide
 * either new custom localId or indicate that the attribute should fall back to the auto-generated localId.
 *
 * @param attribute - attribute to modify
 * @param modifications - modification function
 * @public
 */
export function modifyAttribute(
    attribute: IAttribute,
    modifications: AttributeModifications = identity,
): IAttribute {
    const builder = new AttributeBuilder(attribute);

    return modifications(builder).build();
}
