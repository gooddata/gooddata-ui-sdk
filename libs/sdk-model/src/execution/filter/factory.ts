// (C) 2019-2022 GoodData Corporation
import { invariant } from "ts-invariant";
import isNil from "lodash/isNil.js";
import {
    ComparisonConditionOperator,
    IAbsoluteDateFilter,
    IAttributeElements,
    IMeasureValueFilter,
    INegativeAttributeFilter,
    IPositiveAttributeFilter,
    IRelativeDateFilter,
    RangeConditionOperator,
    IRankingFilter,
    RankingFilterOperator,
} from "./index.js";
import { attributeDisplayFormRef, IAttribute, isAttribute, attributeLocalId } from "../attribute/index.js";
import { Identifier, isObjRef, LocalIdRef, ObjRef, ObjRefInScope } from "../../objRef/index.js";
import { IMeasure, isMeasure, measureLocalId } from "../measure/index.js";
import { idRef, localIdRef } from "../../objRef/factory.js";
import { DateAttributeGranularity } from "../../base/dateGranularities.js";

/**
 * Creates a new positive attribute filter.
 *
 * @remarks
 * NOTE: when specifying attribute element using URIs (primary keys), please keep in mind that they MAY NOT be transferable
 * across workspaces. On some backends (such as bear) same element WILL have different URI in each workspace.
 * In general we recommend using URIs only if your code retrieves them at runtime from backend using elements query
 * or from the data view's headers. Hardcoding URIs is never a good idea, if you find yourself doing that,
 * please consider specifying attribute elements by value
 *
 * @param attributeOrRef - either instance of attribute to create filter for or ref or identifier of attribute's display form
 * @param inValues - values to filter for; these can be either specified as AttributeElements object or as an array
 *  of attribute element _values_; if you specify empty array, then the filter will be noop and will be ignored
 * @public
 */
export function newPositiveAttributeFilter(
    attributeOrRef: IAttribute | ObjRef | Identifier,
    inValues: IAttributeElements | string[],
): IPositiveAttributeFilter {
    const objRef = isObjRef(attributeOrRef)
        ? attributeOrRef
        : typeof attributeOrRef === "string"
        ? idRef(attributeOrRef)
        : attributeDisplayFormRef(attributeOrRef);

    const inObject: IAttributeElements = Array.isArray(inValues) ? { values: inValues } : inValues;

    return {
        positiveAttributeFilter: {
            displayForm: objRef,
            in: inObject,
        },
    };
}

/**
 * Creates a new negative attribute filter.
 *
 * @remarks
 * NOTE: when specifying attribute element using URIs (primary keys), please keep in mind that they MAY NOT be transferable
 * across workspaces. On some backends (such as bear) same element WILL have different URI in each workspace.
 * In general we recommend using URIs only if your code retrieves them at runtime from backend using elements query
 * or from the data view's headers. Hardcoding URIs is never a good idea, if you find yourself doing that,
 * please consider specifying attribute elements by value
 *
 * @param attributeOrRef - either instance of attribute to create filter for or ref or identifier of attribute's display form
 * @param notInValues - values to filter out; these can be either specified as AttributeElements object or as an array
 *  of attribute element _values_; if you specify empty array, then the filter will be noop and will be ignored
 * @public
 */
export function newNegativeAttributeFilter(
    attributeOrRef: IAttribute | ObjRef | Identifier,
    notInValues: IAttributeElements | string[],
): INegativeAttributeFilter {
    const objRef = isObjRef(attributeOrRef)
        ? attributeOrRef
        : typeof attributeOrRef === "string"
        ? idRef(attributeOrRef)
        : attributeDisplayFormRef(attributeOrRef);

    const notInObject: IAttributeElements = Array.isArray(notInValues)
        ? { values: notInValues }
        : notInValues;

    return {
        negativeAttributeFilter: {
            displayForm: objRef,
            notIn: notInObject,
        },
    };
}

/**
 * Creates a new absolute date filter.
 *
 * @param dateDataSet - ref or identifier of the date data set to filter on
 * @param from - start of the interval in ISO-8601 calendar date format
 * @param to - end of the interval in ISO-8601 calendar date format
 * @public
 */
export function newAbsoluteDateFilter(
    dateDataSet: ObjRef | Identifier,
    from: string,
    to: string,
): IAbsoluteDateFilter {
    const dataSet = isObjRef(dateDataSet) ? dateDataSet : idRef(dateDataSet);
    return {
        absoluteDateFilter: {
            dataSet,
            from,
            to,
        },
    };
}

/**
 * Creates a new relative date filter.
 *
 * @param dateDataSet - ref or identifier of the date data set to filter on
 * @param granularity - granularity of the filters (month, year, etc.)
 * @param from - start of the interval – negative numbers mean the past, zero means today, positive numbers mean the future
 * @param to - end of the interval – negative numbers mean the past, zero means today, positive numbers mean the future
 *
 * See also {@link DateAttributeGranularity} and {@link DateGranularity}
 * @public
 */
export function newRelativeDateFilter(
    dateDataSet: ObjRef | Identifier,
    granularity: DateAttributeGranularity,
    from: number,
    to: number,
): IRelativeDateFilter {
    const dataSet = isObjRef(dateDataSet) ? dateDataSet : idRef(dateDataSet);
    return {
        relativeDateFilter: {
            dataSet,
            granularity,
            from,
            to,
        },
    };
}

/**
 * Creates a new all time date filter. This filter is used to indicate that there should be no filtering on the given date data set.
 *
 * @param dateDataSet - ref or identifier of the date data set to filter on
 * @public
 */
export function newAllTimeFilter(dateDataSet: ObjRef | Identifier): IRelativeDateFilter {
    const dataSet = isObjRef(dateDataSet) ? dateDataSet : idRef(dateDataSet);
    return {
        relativeDateFilter: {
            dataSet,
            granularity: "ALL_TIME_GRANULARITY",
            from: 0,
            to: 0,
        },
    };
}

function convertMeasureOrRefToObjRefInScope(measureOrRef: IMeasure | ObjRefInScope | string): ObjRefInScope {
    return isMeasure(measureOrRef)
        ? localIdRef(measureLocalId(measureOrRef))
        : typeof measureOrRef === "string"
        ? localIdRef(measureOrRef)
        : measureOrRef;
}

function convertAttributeOrRefToObjRefInScope(
    attributeOrRef: IAttribute | ObjRefInScope | string,
): ObjRefInScope {
    return isAttribute(attributeOrRef)
        ? localIdRef(attributeLocalId(attributeOrRef))
        : typeof attributeOrRef === "string"
        ? localIdRef(attributeOrRef)
        : attributeOrRef;
}

/**
 * Creates a new measure value filter.
 *
 * @param measureOrRef - instance of measure to filter, or reference of the measure object; if instance of measure is
 *   provided, then it is assumed this measure is in scope of execution and will be referenced by the filter by
 *   its local identifier
 * @param operator - comparison or range operator to use in the filter
 * @param value - the value to compare to
 * @param treatNullValuesAs - value to use instead of null values
 * @public
 */
export function newMeasureValueFilter(
    measureOrRef: IMeasure | ObjRefInScope | string,
    operator: ComparisonConditionOperator,
    value: number,
    treatNullValuesAs?: number,
): IMeasureValueFilter;

/**
 * Creates a new measure value filter.
 *
 * @param measureOrRef - instance of measure to filter, or reference of the measure object; if instance of measure is
 *   provided, then it is assumed this measure is in scope of execution and will be referenced by the filter by
 *   its local identifier
 * @param operator - range operator to use in the filter
 * @param from - the start of the range
 * @param to - the end of the range
 * @param treatNullValuesAs - value to use instead of null values
 * @public
 */
export function newMeasureValueFilter(
    measureOrRef: IMeasure | ObjRefInScope | LocalIdRef | string,
    operator: RangeConditionOperator,
    from: number,
    to: number,
    treatNullValuesAs?: number,
): IMeasureValueFilter;

/**
 * Creates a new measure value filter.
 *
 * @param measureOrRef - instance of measure to filter, or reference of the measure object; if instance of measure is
 *   provided, then it is assumed this measure is in scope of execution and will be referenced by the filter by
 *   its local identifier
 * @param operator - comparison or range operator to use in the filter
 * @param val1 - first numeric value, used as value in comparison or as 'from' value in range operators
 * @param val2OrTreatNullValuesAsInComparison - second numeric value, required in range operators and used in 'to' value; optional in comparison operators used as 'treatNullValuesAs' value
 * @param treatNullValuesAsInRange - third numeric value, optional in range operators and used as 'treatNullValuesAs' value; optional and ignored in comparison operators
 * @public
 */
export function newMeasureValueFilter(
    measureOrRef: IMeasure | ObjRefInScope | string,
    operator: ComparisonConditionOperator | RangeConditionOperator,
    val1: number,
    val2OrTreatNullValuesAsInComparison?: number,
    treatNullValuesAsInRange?: number,
): IMeasureValueFilter {
    const ref = convertMeasureOrRefToObjRefInScope(measureOrRef);

    if (operator === "BETWEEN" || operator === "NOT_BETWEEN") {
        invariant(
            val2OrTreatNullValuesAsInComparison !== undefined,
            "measure value filter with range operator requires two numeric values",
        );

        const nullValuesProp = !isNil(treatNullValuesAsInRange)
            ? { treatNullValuesAs: treatNullValuesAsInRange }
            : {};

        return {
            measureValueFilter: {
                measure: ref,
                condition: {
                    range: {
                        operator,
                        from: val1,
                        to: val2OrTreatNullValuesAsInComparison!,
                        ...nullValuesProp,
                    },
                },
            },
        };
    } else {
        const nullValuesProp = !isNil(val2OrTreatNullValuesAsInComparison)
            ? { treatNullValuesAs: val2OrTreatNullValuesAsInComparison }
            : {};

        return {
            measureValueFilter: {
                measure: ref,
                condition: {
                    comparison: {
                        operator,
                        value: val1,
                        ...nullValuesProp,
                    },
                },
            },
        };
    }
}

/**
 * Creates a new ranking filter.
 *
 * @param measureOrRef - instance of measure to filter, or reference of the measure object; if instance of measure is
 *   provided, then it is assumed this measure is in scope of execution and will be referenced by the filter by
 *   its local identifier
 * @param operator - TOP or BOTTOM operator to use in the filter
 * @param value - Number of values to use in filter
 * @param attributesOrRefs - Array of attributes used in filter, or reference of the attribute object. If instance of attribute is
 *   provided, then it is assumed this attribute is in scope of execution and will be referenced by the filter by
 *   its local identifier
 * @public
 */
export function newRankingFilter(
    measureOrRef: IMeasure | ObjRefInScope | string,
    attributesOrRefs: Array<IAttribute | ObjRefInScope | string>,
    operator: RankingFilterOperator,
    value: number,
): IRankingFilter;

/**
 * Creates a new ranking filter.
 *
 * @param measureOrRef - instance of measure to filter, or reference of the measure object; if instance of measure is
 *   provided, then it is assumed this measure is in scope of execution and will be referenced by the filter by
 *   its local identifier
 * @param operator - TOP or BOTTOM operator to use in the filter
 * @param value - Number of values to use in filter
 * @public
 */
export function newRankingFilter(
    measureOrRef: IMeasure | ObjRefInScope | string,
    operator: RankingFilterOperator,
    value: number,
): IRankingFilter;

/**
 * Creates a new ranking filter.
 *
 * @param measureOrRef - instance of measure to filter, or reference of the measure object; if instance of measure is
 *   provided, then it is assumed this measure is in scope of execution and will be referenced by the filter by
 *   its local identifier
 * @param attributesOrRefsOrOperator - Array of attributes used in filter, or reference of the attribute object. If instance of attribute is
 *   provided, then it is assumed this attribute is in scope of execution and will be referenced by the filter by
 *   its local identifier. If attributes are not specified it's TOP or BOTTOM operator to use in the filter
 * @param operatorOrValue - Number of values to use in filter or operator if attributes are not speciied
 * @param valueOrNothing - Value or nothing if attributes are not specified
 * @public
 */
export function newRankingFilter(
    measureOrRef: IMeasure | ObjRefInScope | string,
    attributesOrRefsOrOperator: Array<IAttribute | ObjRefInScope | string> | RankingFilterOperator,
    operatorOrValue: RankingFilterOperator | number,
    valueOrNothing?: number,
): IRankingFilter {
    if (typeof attributesOrRefsOrOperator === "string" && typeof operatorOrValue === "number") {
        const measureRef = convertMeasureOrRefToObjRefInScope(measureOrRef);
        return {
            rankingFilter: {
                measure: measureRef,
                operator: attributesOrRefsOrOperator,
                value: operatorOrValue,
            },
        };
    } else if (typeof operatorOrValue === "string" && valueOrNothing) {
        const measureRef = convertMeasureOrRefToObjRefInScope(measureOrRef);
        const attributeRefs = (attributesOrRefsOrOperator as Array<IAttribute | ObjRefInScope | string>).map(
            convertAttributeOrRefToObjRefInScope,
        );
        const attributesProp = attributeRefs.length ? { attributes: attributeRefs } : {};
        return {
            rankingFilter: {
                measure: measureRef,
                operator: operatorOrValue,
                value: valueOrNothing,
                ...attributesProp,
            },
        };
    }

    throw new Error("Ranking filter requires measure, operator and value");
}
