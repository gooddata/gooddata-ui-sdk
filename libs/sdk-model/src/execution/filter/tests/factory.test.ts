// (C) 2019-2020 GoodData Corporation
import { Account, Department, Won } from "../../../../__mocks__/model.js";
import {
    newAbsoluteDateFilter,
    newMeasureValueFilter,
    newNegativeAttributeFilter,
    newPositiveAttributeFilter,
    newRelativeDateFilter,
    newRankingFilter,
} from "../factory.js";
import { localIdRef, idRef } from "../../../objRef/factory.js";
import { IAttribute, attributeIdentifier, IdentifierRef, LocalIdRef } from "../../../index.js";
import { attributeLocalId } from "../../attribute/index.js";

describe("filter factory", () => {
    describe("newPositiveAttributeFilter", () => {
        it("should generate correct uri filter", () => {
            expect(newPositiveAttributeFilter("foo", { uris: ["bar", "baz"] })).toMatchSnapshot();
        });

        it("should generate correct value filter", () => {
            expect(newPositiveAttributeFilter("foo", { values: ["bar", "baz"] })).toMatchSnapshot();
        });

        it("should generate correct value filter when input is array of values", () => {
            expect(newPositiveAttributeFilter("foo", ["bar", "baz"])).toMatchSnapshot();
        });

        it("should generate correct filter attribute object is on input", () => {
            expect(newPositiveAttributeFilter(Account.Name, ["bar", "baz"])).toMatchSnapshot();
        });
    });

    describe("newNegativeAttributeFilter", () => {
        it("should generate correct filter", () => {
            expect(newNegativeAttributeFilter("foo", { uris: ["bar", "baz"] })).toMatchSnapshot();
        });
        it("should generate correct textual filter", () => {
            expect(newNegativeAttributeFilter("foo", { values: ["bar", "baz"] })).toMatchSnapshot();
        });
        it("should generate correct textual filter when input is array of values", () => {
            expect(newNegativeAttributeFilter("foo", ["bar", "baz"])).toMatchSnapshot();
        });
        it("should generate correct filter when input is IAttribute instance", () => {
            expect(newNegativeAttributeFilter(Account.Name, ["bar", "baz"])).toMatchSnapshot();
        });
    });

    describe("newAbsoluteDateFilter", () => {
        it("should generate correct filter", () => {
            expect(newAbsoluteDateFilter("foo", "2018-01-01", "2018-12-31")).toMatchSnapshot();
        });
    });

    describe("newRelativeDateFilter", () => {
        it("should generate correct filter", () => {
            expect(newRelativeDateFilter("foo", "GDC.time.month", 1, 3)).toMatchSnapshot();
        });
    });

    describe("newMeasureValueFilter", () => {
        it("should generate comparison filter for measure object", () => {
            expect(newMeasureValueFilter(Won, "EQUAL_TO", 11)).toMatchSnapshot();
        });
        it("should generate comparison filter for measure identifier", () => {
            expect(newMeasureValueFilter(idRef("identifier"), "EQUAL_TO", 11)).toMatchSnapshot();
        });
        it("should generate comparison filter for measure local identifier", () => {
            expect(newMeasureValueFilter(localIdRef("measureObjLocalId"), "EQUAL_TO", 11)).toMatchSnapshot();
        });
        it("should generate comparison filter for measure object with treatNullValuesAs", () => {
            expect(newMeasureValueFilter(Won, "EQUAL_TO", 11, 42)).toMatchSnapshot();
        });
        it("should generate range filter for measure object", () => {
            expect(newMeasureValueFilter(Won, "BETWEEN", 0, 100)).toMatchSnapshot();
        });
        it("should generate ranger filter for measure identifier", () => {
            expect(
                newMeasureValueFilter(localIdRef("measureObjLocalId"), "BETWEEN", 0, 100),
            ).toMatchSnapshot();
        });
        it("should generate ranger filter for measure local identifier", () => {
            expect(newMeasureValueFilter("measureObjLocalId", "BETWEEN", 0, 100)).toMatchSnapshot();
        });
        it("should generate range filter for measure object with treatNullValuesAs", () => {
            expect(newMeasureValueFilter(Won, "BETWEEN", 0, 100, 42)).toMatchSnapshot();
        });
    });

    describe("newRankingFilter", () => {
        it("should generate top filter for measure object", () => {
            expect(newRankingFilter(Won, "TOP", 10)).toMatchSnapshot();
        });
        it("should generate top filter for measure identifier", () => {
            expect(newRankingFilter(idRef("identifier"), "BOTTOM", 30)).toMatchSnapshot();
        });
        it("should generate top filter for measure local identifier ref", () => {
            expect(newRankingFilter(localIdRef("measureObjLocalId"), "TOP", 5)).toMatchSnapshot();
        });
        it("should generate top filter for measure local identifier", () => {
            expect(newRankingFilter("measureObjLocalId", "TOP", 5)).toMatchSnapshot();
        });
        it("should generate bottom filter for measure object", () => {
            expect(newRankingFilter(Won, "BOTTOM", 10)).toMatchSnapshot();
        });
        it("should remove empty array of attributes as it's invalid state", () => {
            const attributes: IAttribute[] = [];
            expect(newRankingFilter(Won, attributes, "BOTTOM", 20)).toMatchSnapshot();
        });
        it("should generate bottom filter for measure object and attribute objects", () => {
            const attributes: IAttribute[] = [Department, Account.Name];
            expect(newRankingFilter(Won, attributes, "BOTTOM", 20)).toMatchSnapshot();
        });
        it("should generate bottom filter for measure object and attribute identifiers", () => {
            const attributes: IdentifierRef[] = [
                idRef(attributeIdentifier(Department) ?? ""),
                idRef(attributeIdentifier(Account.Default) ?? ""),
            ];
            expect(newRankingFilter(Won, attributes, "BOTTOM", 5)).toMatchSnapshot();
        });
        it("should generate bottom filter for measure object and attribute local identifiers refs", () => {
            const attributes: LocalIdRef[] = [
                localIdRef(attributeLocalId(Department)),
                localIdRef(attributeLocalId(Account.Default)),
            ];
            expect(newRankingFilter(Won, attributes, "BOTTOM", 3)).toMatchSnapshot();
        });
        it("should generate bottom filter for measure object and attribute local identifiers", () => {
            const attributes: string[] = [attributeLocalId(Department), attributeLocalId(Account.Default)];
            expect(newRankingFilter(Won, attributes, "BOTTOM", 3)).toMatchSnapshot();
        });
    });
});
