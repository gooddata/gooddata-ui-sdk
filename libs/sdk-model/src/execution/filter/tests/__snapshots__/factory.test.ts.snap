// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`filter factory > newAbsoluteDateFilter > should generate correct filter 1`] = `
{
  "absoluteDateFilter": {
    "dataSet": {
      "identifier": "foo",
    },
    "from": "2018-01-01",
    "localIdentifier": "testLocalId",
    "to": "2018-12-31",
  },
}
`;

exports[`filter factory > newMeasureValueFilter > should generate comparison filter for measure identifier 1`] = `
{
  "measureValueFilter": {
    "condition": {
      "comparison": {
        "operator": "EQUAL_TO",
        "value": 11,
      },
    },
    "measure": {
      "identifier": "identifier",
    },
  },
}
`;

exports[`filter factory > newMeasureValueFilter > should generate comparison filter for measure local identifier 1`] = `
{
  "measureValueFilter": {
    "condition": {
      "comparison": {
        "operator": "EQUAL_TO",
        "value": 11,
      },
    },
    "measure": {
      "localIdentifier": "measureObjLocalId",
    },
  },
}
`;

exports[`filter factory > newMeasureValueFilter > should generate comparison filter for measure object 1`] = `
{
  "measureValueFilter": {
    "condition": {
      "comparison": {
        "operator": "EQUAL_TO",
        "value": 11,
      },
    },
    "measure": {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
  },
}
`;

exports[`filter factory > newMeasureValueFilter > should generate comparison filter for measure object with treatNullValuesAs 1`] = `
{
  "measureValueFilter": {
    "condition": {
      "comparison": {
        "operator": "EQUAL_TO",
        "treatNullValuesAs": 42,
        "value": 11,
      },
    },
    "measure": {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
  },
}
`;

exports[`filter factory > newMeasureValueFilter > should generate range filter for measure object 1`] = `
{
  "measureValueFilter": {
    "condition": {
      "range": {
        "from": 0,
        "operator": "BETWEEN",
        "to": 100,
      },
    },
    "measure": {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
  },
}
`;

exports[`filter factory > newMeasureValueFilter > should generate range filter for measure object with treatNullValuesAs 1`] = `
{
  "measureValueFilter": {
    "condition": {
      "range": {
        "from": 0,
        "operator": "BETWEEN",
        "to": 100,
        "treatNullValuesAs": 42,
      },
    },
    "measure": {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
  },
}
`;

exports[`filter factory > newMeasureValueFilter > should generate ranger filter for measure identifier 1`] = `
{
  "measureValueFilter": {
    "condition": {
      "range": {
        "from": 0,
        "operator": "BETWEEN",
        "to": 100,
      },
    },
    "measure": {
      "localIdentifier": "measureObjLocalId",
    },
  },
}
`;

exports[`filter factory > newMeasureValueFilter > should generate ranger filter for measure local identifier 1`] = `
{
  "measureValueFilter": {
    "condition": {
      "range": {
        "from": 0,
        "operator": "BETWEEN",
        "to": 100,
      },
    },
    "measure": {
      "localIdentifier": "measureObjLocalId",
    },
  },
}
`;

exports[`filter factory > newNegativeAttributeFilter > should generate correct filter 1`] = `
{
  "negativeAttributeFilter": {
    "displayForm": {
      "identifier": "foo",
    },
    "notIn": {
      "uris": [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory > newNegativeAttributeFilter > should generate correct filter when input is IAttribute instance 1`] = `
{
  "negativeAttributeFilter": {
    "displayForm": {
      "identifier": "label.account.id.name",
      "type": "displayForm",
    },
    "notIn": {
      "values": [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory > newNegativeAttributeFilter > should generate correct textual filter 1`] = `
{
  "negativeAttributeFilter": {
    "displayForm": {
      "identifier": "foo",
    },
    "notIn": {
      "values": [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory > newNegativeAttributeFilter > should generate correct textual filter when input is array of values 1`] = `
{
  "negativeAttributeFilter": {
    "displayForm": {
      "identifier": "foo",
    },
    "notIn": {
      "values": [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory > newPositiveAttributeFilter > should generate correct filter attribute object is on input 1`] = `
{
  "positiveAttributeFilter": {
    "displayForm": {
      "identifier": "label.account.id.name",
      "type": "displayForm",
    },
    "in": {
      "values": [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory > newPositiveAttributeFilter > should generate correct uri filter 1`] = `
{
  "positiveAttributeFilter": {
    "displayForm": {
      "identifier": "foo",
    },
    "in": {
      "uris": [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory > newPositiveAttributeFilter > should generate correct value filter 1`] = `
{
  "positiveAttributeFilter": {
    "displayForm": {
      "identifier": "foo",
    },
    "in": {
      "values": [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory > newPositiveAttributeFilter > should generate correct value filter when input is array of values 1`] = `
{
  "positiveAttributeFilter": {
    "displayForm": {
      "identifier": "foo",
    },
    "in": {
      "values": [
        "bar",
        "baz",
      ],
    },
  },
}
`;

exports[`filter factory > newRankingFilter > should generate bottom filter for measure object 1`] = `
{
  "rankingFilter": {
    "measure": {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
    "operator": "BOTTOM",
    "value": 10,
  },
}
`;

exports[`filter factory > newRankingFilter > should generate bottom filter for measure object and attribute identifiers 1`] = `
{
  "rankingFilter": {
    "attributes": [
      {
        "identifier": "label.owner.department",
      },
      {
        "identifier": "label.account.id",
      },
    ],
    "measure": {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
    "operator": "BOTTOM",
    "value": 5,
  },
}
`;

exports[`filter factory > newRankingFilter > should generate bottom filter for measure object and attribute local identifiers 1`] = `
{
  "rankingFilter": {
    "attributes": [
      {
        "localIdentifier": "a_label.owner.department",
      },
      {
        "localIdentifier": "a_label.account.id",
      },
    ],
    "measure": {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
    "operator": "BOTTOM",
    "value": 3,
  },
}
`;

exports[`filter factory > newRankingFilter > should generate bottom filter for measure object and attribute local identifiers refs 1`] = `
{
  "rankingFilter": {
    "attributes": [
      {
        "localIdentifier": "a_label.owner.department",
      },
      {
        "localIdentifier": "a_label.account.id",
      },
    ],
    "measure": {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
    "operator": "BOTTOM",
    "value": 3,
  },
}
`;

exports[`filter factory > newRankingFilter > should generate bottom filter for measure object and attribute objects 1`] = `
{
  "rankingFilter": {
    "attributes": [
      {
        "localIdentifier": "a_label.owner.department",
      },
      {
        "localIdentifier": "a_label.account.id.name",
      },
    ],
    "measure": {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
    "operator": "BOTTOM",
    "value": 20,
  },
}
`;

exports[`filter factory > newRankingFilter > should generate top filter for measure identifier 1`] = `
{
  "rankingFilter": {
    "measure": {
      "identifier": "identifier",
    },
    "operator": "BOTTOM",
    "value": 30,
  },
}
`;

exports[`filter factory > newRankingFilter > should generate top filter for measure local identifier 1`] = `
{
  "rankingFilter": {
    "measure": {
      "localIdentifier": "measureObjLocalId",
    },
    "operator": "TOP",
    "value": 5,
  },
}
`;

exports[`filter factory > newRankingFilter > should generate top filter for measure local identifier ref 1`] = `
{
  "rankingFilter": {
    "measure": {
      "localIdentifier": "measureObjLocalId",
    },
    "operator": "TOP",
    "value": 5,
  },
}
`;

exports[`filter factory > newRankingFilter > should generate top filter for measure object 1`] = `
{
  "rankingFilter": {
    "measure": {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
    "operator": "TOP",
    "value": 10,
  },
}
`;

exports[`filter factory > newRankingFilter > should remove empty array of attributes as it's invalid state 1`] = `
{
  "rankingFilter": {
    "measure": {
      "localIdentifier": "m_afSEwRwdbMeQ",
    },
    "operator": "BOTTOM",
    "value": 20,
  },
}
`;

exports[`filter factory > newRelativeDateFilter > should generate correct filter 1`] = `
{
  "relativeDateFilter": {
    "dataSet": {
      "identifier": "foo",
    },
    "from": 1,
    "granularity": "GDC.time.month",
    "localIdentifier": "testLocalId",
    "to": 3,
  },
}
`;
